mako/utils/transformers.cc

```diff
@@ -0,0 +1,114 @@
+// Adapted from https://github.com/vllm-project/vllm/blob/v0.2.7/vllm/model_executor/weight_utils.py
+// Copyright 2024 The Mako Authors
+// Copyright 2023 The vLLM team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "mako/utils/transformers.h"
+
+#include <vector>
+#include <iostream>
+#include <fstream>
+#include <tuple>
+#include <string>
+#include <filesystem>
+#include <algorithm>
+
+namespace fs = std::filesystem;
+
+/// \brief Utility to find weight files from model directory
+/// \param model_name_or_path A path to a directory containing model weights saved using
+/// \param
+/// \param
+/// \param
+/// \param
+/// \return
+std::tuple<std::string, std::vector<std::string>, bool> prepare_hf_model_weights (
```

Comment:
Please use ``static inline`` primitives before static functions. Also, remove black space between function name and bracket.
i.e., ``static inline std::tuple<std::string, std::vector<std::string>, bool> prepare_hf_model_weights(``

Reply:
First of all, thanks for your comment @9rum!

Using `static` would be great. I'll edit my code to use static.
But using `inline` don't look good to me.

reasons
1. We only use `prepare_hf_model_weights` once when we get model weights' file's path and it is not just simple calculation code(It contains 100s of lines). Using inline makes code larger and might cause thrashing problem.
2. `snapshot_download` that I planned to implement have large I/O works(it download model from huggingface) and using inline function can cause blocking behavior.

It don't cause serious performance problem now, but I think using inline function can be a future-obstacles.

Here is my reference of my thoughts. Feel free to comment your opinion:)
- [drdobbs](https://www.drdobbs.com/to-inline-or-not-to-inline/184405660)
- [isocpp.org](https://isocpp.org/wiki/faq/inline-functions#overview-inline-fns)

Reply:
I think you may have some misunderstandings about inline functions.
Inlining is a common optimization strategy for reducing ``jmp`` and ``mov`` instructions, by copying the entire inlined code to the caller function. Thus **inlining is not related to thrashing or blocking problem**.
As mentioned in your references, such code copying may reduce a few instruction calls, while making the generated executable size larger (i.e., there is a trade-off between program's speed and size).
However, the executable size does not matter much in the modern HPC environments, so I suggest to inline the function for better speedup.